Add the UI-only autorenewal toggle

southp opened the pull request:
#### Changes proposed in this Pull Request

*This PR is part of the attempt of adding a self-serving plan subscription autorenewal toggle. For more details, please refer to p2-p9jf6J-1GZ*

This PR adds a UI-only autorenewal toggle under the purchase management page, according to @nodeone 's design:
![demo-renew-toggle](https://user-images.githubusercontent.com/1842898/57691287-64bf7c00-7676-11e9-9c4e-f1a0a6bf75ce.gif)

It is only enabled in the development environment through `autorenewal-toggle` feature flag. UI-only means that it does nothing for now. The actual functionalities will be wired in through the future PRs.

#### Testing instructions

Note that this is UI-only and is only enabled in the development environment.

* Log in as a user who has plan subscriptions and non-plan purchases
* Go to http://calypso.localhost:3000/me/purchases/
* Click on any plan subscriptions, see the toggle, and you should be able to toggle it as expected.
* Click on any non-plan purchases and there shouldn't be this toggle.

michelleweber commented:
What is the copy that needs reviewing here? The toggle itself looks fine.

southp commented:
@michelleweber Thanks for the fast check-in. You are right, it's pretty small and probably doesn't need much copy review:

* Auto-renew is ON / OFF
* You will be billed on yyyy/mm/dd
* Expires on yyyy/mm/dd

I guess I did that in the first place because I've developed the habit of labeling copy review whenever I touch any text, so it's kind of a reflex action 😅

spen commented on client/me/purchases/manage-purchase/purchase-meta.jsx:
```diff
@@ -56,6 +58,17 @@ class PurchaseMeta extends Component {
+
+		if ( config.isEnabled( 'autorenewal-toggle' ) ) {
+			// TODO: remove this once the proper state has been introduced.
+			this.state = {
```
I'm a little hesitant with this approach as we can end up with or without `this.state` here, depending on the flag.
Of course, we only access `this.state` in logic that's guarded with the flag check, but it still makes me wonder if we can avoid potential future bugs.

Something like this comes to mind:

```js
state = {
  ...config.isEnabled( 'autorenewal-toggle' ) && {
    isAutorenewalEnabled: false,
  },
};
```

That way, `this.state` is always set to at least an empty object and future diffs can leave a much smaller change footprint:

```js
state = {
  someStateFromTheFuture: true,
  ...config.isEnabled( 'autorenewal-toggle' ) && {
    isAutorenewalEnabled: false,
  },
};
```

southp replied:
Great advice! Added in 92fc8e4.

southp commented on client/me/purchases/manage-purchase/purchase-meta.jsx:
```diff
@@ -287,12 +300,55 @@ class PurchaseMeta extends Component {
+							renewDate: purchase.renewMoment.format( 'LL' ),
+						},
+				  } )
+				: translate( 'Expires on %(expireDate)s', {
```
Ah, nice catch. It wouldn't make any sense for expired subscriptions to turn on the auto-renewal, so I've excluded them in d987cf7. Now they will look just like what we have now:
<img width="759" alt="螢幕快照 2019-05-17 下午3 17 05" src="https://user-images.githubusercontent.com/1842898/57910313-95422880-78b7-11e9-9edc-9a01fd0afab8.png">

southp commented:
@niranjan-uma-shankar Good catch. It will be rolled out to G Suite in the future, but not now. I have exclude them in 3cf639b.

southp commented:
@niranjan-uma-shankar @spen Would this UI tweak work for you?
![demo-quarterly-divided](https://user-images.githubusercontent.com/1842898/57910414-d3d7e300-78b7-11e9-8854-008bbd57879e.gif)

What I did in 0dcde1f 686f240 and were:
* Move the toggle down to the bottom.
* Divide the four into four equal, fixed width columns so they won't flick with the change of text.

niranjan-uma-shankar commented:
Couple of observations:

1. When I tried this on a new test account, it works as described. However, when I tried it on my personal WPCOM account, clicking the toggle resulted in a blank screen: https://cloudup.com/cIDx-jAfdwo

The console gave the error:

```
purchase-meta.jsx:334 Uncaught TypeError: Cannot read property 'format' of null
    at PurchaseMeta.renderExpiration (purchase-meta.jsx:334)
    at PurchaseMeta.render (purchase-meta.jsx:389)
    at finishClassComponent (react-dom.development.js:14741)
    at updateClassComponent (react-dom.development.js:14696)
    at beginWork (react-dom.development.js:15644)
    at performUnitOfWork (react-dom.development.js:19312)
    at workLoop (react-dom.development.js:19352)
    at HTMLUnknownElement.callCallback (react-dom.development.js:149)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:199)
    at invokeGuardedCallback (react-dom.development.js:256)
```

I checked and noticed that the `renewMoment ` property takes a `null` value for my plan purchase. The only thing I can think of is different on my account is that I added a Premium upgrade to my account through the Store admin page. Do such upgrades give a null value for renewMoment?

If we do occasionally upgrade users via store admin, then this issue is worth a fix. Thoughts?

2. Not a strong opinion, but I think it would be better if the whole date breaks into a new line. For eg., I presently see this:

<img width="164" alt="Screenshot 2019-05-20 at 10 35 41 AM" src="https://user-images.githubusercontent.com/1269602/57998180-feab7c80-7aed-11e9-9ff0-5a10b8ec9a7a.png">

The date `9` is awkwardly sticking to the end of the line, and it could be easily missed.

I'm thinking it could be better if wrap the entire date in a <span> and add a `white-space: nowrap` style property to it:
`<span style="white-space:nowrap">9 April 2020</span>` makes it look like this:

<img width="174" alt="Screenshot 2019-05-20 at 10 36 25 AM" src="https://user-images.githubusercontent.com/1269602/57998257-4fbb7080-7aee-11e9-93f6-21e20f33f42d.png">

southp commented:
Great catch and wonderful suggestion on `nowrap`, @niranjan-uma-shankar !
I've addressed your feedbacks in 5a298c7 and 977440e, so it shouldn't crash for your test account anymore and the date string should stay on the same line. Please have a revisit to see if it works for you when you have some time 🙇

southp commented:
@spen Ah! Thanks for the tip. I forgot we can use property initializer nowadays. Added in 84c3787  :)
